content/post/postgres/rfc-extension-packaging-lookup.md

```diff
@@ -0,0 +1,496 @@
+`/opt/pg17/ext/pair/pair.control`, then it will load files only from the
+appropriate subdirectories, e.g.:
+
+*   SQL files from `/opt/pg17/ext/pair/share`
+*   Shared module files from `/opt/pg17/ext/pair/lib`
+
+### PGXS
+
+Update the extension installation behavior of [PGXS] to install extension
+files into the new locations. A new variable, `EXTDIR`, will define the
+directory into which to install an extension, and will default to
+`--extdir-site`. It can be set to the values `$extdir_site`, `$extdir_vendor`,
+or `$extdir_core`, or to any literal path.
+
+The `$EXTENSION` variable will be changed to allow only one extension name. If
+it's set, the installation behavior will be changed for the following
+variables:
+
+*   `EXTENSION`: Creates `$EXTDIR/$EXTENSION`, installs
+    `$EXTDIR/$EXTENSION/$EXTENSION.control`
+*   `MODULES` and `MODULE_big`: Installed into `$EXTDIR/$EXTENSION/lib`
+*   `MODULEDIR`: Removed
+*   `DATA`: Installed into `$EXTDIR/$EXTENSION/share`
+*   `DATA_built`: Installed into `$EXTDIR/$EXTENSION/share`
+*   `DATA_TSEARCH`: Installed into `$EXTDIR/$EXTENSION/share/tsearch_data`
+*   `DOCS`: Installed into `$EXTDIR/$EXTENSION/doc`
+*   `PROGRAM`, `SCRIPTS` and `SCRIPTS_built`: Installed into
+    `$EXTDIR/$EXTENSION/bin`
+
+Another new variable, `LINKBINS`, will default to true and symlink
+`$EXTDIR/$EXTENSION/bin` files in `pg_config --bindir`. Installers can set it
+to false to skip the symlinking, e.g., for immutable Postgres installs.
+
+> [!NOTE] External projects that install extensions without using PGXS, like
+> [pgrx], must also be updated to either follow the same pattern or to
+> delegate installation to [PGXS].
+
+### MODULE_PATHNAME
+
+Update the installer to replace `MODULE_PATHNAME` in SQL scripts with the new
+install path for shared modules, `$EXTDIR/$EXTENSION/lib`.
+
+### Control File
+
+The `directory` and `module_pathname` control file variables will be
+deprecated and ignored.
+
+## Use Cases
+
+Here’s how the proposed file layout and `extension_path` GUC address the [use
+cases](#the-problem) that inspired this RFC.
+
+### Packager Testing
+
+A packager who wants to run tests without modifying a PostgreSQL install would
+follow these steps:
+
+*   Set the `extension_path` GUC to search the site extension directory under
+    the packaging install. Something like
+    `$RPM_BUILD_ROOT/$(pg_config --extdir-vendor)`
+*   Install the extension into that directory:
+    `make install EXTDIR=$RPM_BUILD_ROOT`
+*   Run `make installcheck`
+
+This will allow PostgreSQL to find and load the extension during the tests.
+The Postgres installation will not have been modified; only the
+`extension_path` will have changed.
+
+### Docker/Kubernetes
+
+To allow extensions to be added to a Docker container and to persist beyond
+its lifetime, one or more [volumes] could be used. A couple of options:
+
+*   Mount the `--extdir-site` and/or `--extdir-vendor` directories as
```

Comment:
This approach works for single node. For multiple nodes you need to repeat the operation for all nodes. While doable (e.g. this is what StackGres does) it's quite a bit of an effort, and may be mentioned here.

Reply:
I assume you use a Kubernetes manifest that defines the shared volume and all your pods use it, yes? I must be misunderstanding, though; what's the extra effort?

Reply:
No, volumes are not shared. If so, that would require the use of a distributed filesystem (e.g. NFS) which almost nobody does (for known reasons).

Most of the setups I know (and definitely ours in StackGres) have one volume per pod. Those volumes are different, and PGDATA is obvously replicated via streaming replication. So anything else you put on top (e.g. extension's content unpackage from elsewhere) needs to be done per volume (repeated).

Which also means that if, for example, a pod dies and it's re-created, the same operation needs to happen again.

And this also means that if a new extension is added or one is removed, the changes required need to be applied in all volumes again!

In StackGres we wrote a ["cluster controller"](https://speakerdeck.com/ongres/postgres-extensions-in-kubernetes?slide=16), a sidecar that implements all this logic.

Reply:
Yes.

I think it still requires the extensions to be placed on a unique directory per extension (as per this RFC) because I presume you cannot mount images in overlapping paths.

I'm also afraid this mechanism works at startup, so changing the extensions would require a pod restart.

But it's still much simpler than the "cluster controller" (though this mechanism works now and the image volume one possibly only with newer Postgres versions and newer K8s versions).

Reply:
I'm gonna spend some time on [Peter's patch](https://postgr.es/m/0d384836-7e6e-4932-af3b-8dad1f6fee43@eisentraut.org) and think about implications for the future. Perhaps we limit it to the control file, SQL files, and shared modules. It still complicates things a bit for extensions that _also_ ship with scripts/CLIs/whatever, but maybe that's a problem for another time.